src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,717 @@
+      _sampler->enroll();
+    }
+    return;
+  }
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate /* 0 */, auto_adapt);
+    _sampler->disenroll();
+  }
+}
+
+void JfrCPUTimeThreadSampling::set_rate(double rate, bool auto_adapt) {
+  assert(rate >= 0, "invariant");
+  if (_instance == nullptr) {
+    return;
+  }
+  instance().set_rate_value(rate, auto_adapt);
+}
+
+void JfrCPUTimeThreadSampling::set_rate_value(double rate, bool auto_adapt) {
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate, auto_adapt);
+  }
+  update_run_state(rate, auto_adapt);
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_create(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::trigger_async_processing_of_cpu_time_jfr_requests() {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->trigger_async_processing_of_cpu_time_jfr_requests();
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  _instance->handle_timer_signal(info, context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(siginfo_t* info, void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+  _sampler->handle_timer_signal(info, context);
+  Atomic::dec(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+}
+
+void JfrCPUTimeThreadSampler::sample_thread(JfrSampleRequest& request, void* ucontext, JavaThread* jt, JfrThreadLocal* tl) {
+  JfrSampleRequestBuilder::build_cpu_time_sample_request(request, ucontext, jt, jt->jfr_thread_local());
+}
+
+static bool check_state(JavaThread* thread) {
+  switch (thread->thread_state()) {
+    case _thread_in_Java:
+    case _thread_in_native:
+      return true;
+    default:
+      return false;
+  }
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(siginfo_t* info, void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrThreadLocal* tl = jt->jfr_thread_local();
+  JfrCPUTimeTraceQueue& queue = tl->cpu_time_jfr_queue();
+  if (!check_state(jt)) {
+      queue.increment_lost_samples();
+    return;
+  }
+  if (!tl->acquire_cpu_time_jfr_enqueue_lock()) {
+    queue.increment_lost_samples();
+    return;
+  }
+
+  JfrCPUTimeSampleRequest request;
+  // the sampling period might be too low for the current Linux configuration
+  // so samples might be skipped and we have to compute the actual period
+  int64_t period = get_sampling_period() * (info->si_overrun + 1);
```

Comment:
Does this calculation have to be done on every signal, by every thread? It seems like something that could be precalculated when the period is set?

Reply:
This might change dynamically, so probably no. Only caching would work, but this is a small optimization for later.